pinot-core/src/main/java/org/apache/pinot/server/realtime/ControllerLeaderLocator.java

```diff
@@ -80,75 +87,127 @@ public static ControllerLeaderLocator getInstance() {
+        return;
+      }
+      for (String partitionName : partitionNames) {
+        int partitionId = LeadControllerUtils.extractPartitionId(partitionName);
+        Map<String, String> partitionStateMap = leadControllerResourceExternalView.getStateMap(partitionName);
+        boolean masterFound = false;
+        // Get master host from partition map. Return null if no master found.
+        for (Map.Entry<String, String> entry : partitionStateMap.entrySet()) {
+          if (MasterSlaveSMD.States.MASTER.name().equals(entry.getValue())) {
+            // Found the controller in master state.
+            // Converts participant id (with Prefix "Controller_") to controller id and assigns it as the leader,
+            // since realtime segment completion protocol doesn't need the prefix in controller instance id.
+            String participantInstanceId = entry.getKey();
+            String controllerInstanceId = LeadControllerUtils.extractControllerInstanceId(participantInstanceId);
+            Pair<String, Integer> leadControllerPair = convertToHostAndPortPair(controllerInstanceId);
+            masterFound = true;
+            _cachedControllerLeaderMap.put(partitionId, leadControllerPair);
+          }
+        }
+        if (!masterFound) {
+          // It's ok to log a warning since we can be in this state for some small time during the migration.
+          // Otherwise, we are attempted to mark this as an error.
+          LOGGER.warn("There is no controller in MASTER state for partition: {} in {}", partitionName,
+              Helix.LEAD_CONTROLLER_RESOURCE_NAME);
+          return;
+        }
+      }
+      LOGGER.info("Refreshed controller leader map successfully.");
+      refreshSucceeded = true;
     } catch (Exception e) {
-      LOGGER.warn("Could not get whether lead controller resource is enabled or not.", e);
-      return false;
+      LOGGER.warn("Caught exception when getting lead controller instance Id from external view of {}",
+          Helix.LEAD_CONTROLLER_RESOURCE_NAME, e);
+    } finally {
+      _cachedControllerLeaderValid = refreshSucceeded;
     }
   }
 
   /**
-   * Gets leader from lead controller resource. Null if there is no leader.
-   * @param rawTableName raw table name.
-   * @return pair of instance hostname and port of Helix cluster leader, e.g. {localhost, 9000}.
+   * Updates lead controller pairs from Helix cluster leader.
    */
-  private Pair<String, Integer> getLeaderFromLeadControllerResource(String rawTableName) {
-    Pair<String, Integer> leaderHostAndPortPair = getLeadControllerInstanceIdForTable(rawTableName);
-    if (leaderHostAndPortPair != null) {
-      return leaderHostAndPortPair;
-    } else {
-      LOGGER.warn("Could not locate leader for table: {}", rawTableName);
-      return null;
+  private void refreshControllerLeaderMapFromHelixClusterLeader() {
+    boolean refreshSucceeded = false;
```

Comment:
Same here, no need to maintain this extra variable

Reply:
Done.